File: server/services/timeline-event.ts
/**
 * Timeline-event authoring service (US-002).
 *
 * Pure helpers that turn an event/period authored on the temporal axis into a
 * reviewable contribution. Nothing here touches fs / express / the
 * source-of-truth TSVs — the authored entry lands in the *contribution queue*
 * (via `ContributionService.submit`) with provenance
 * `entityData.source = 'user-authored'`, so a human reviewer promotes it into
 * `data/source/lexicons/culture-events.tsv` later (US-009).
 *
 * Two entry kinds are supported (matching the acceptance criteria — "events,
 * period markers, and date ranges"):
 *   - `event`  — a point-in-time marker at a single year (`timePeriodStart`).
 *   - `period` — a dated range (`timePeriodStart`..`timePeriodEnd`).
 *
 * `serializeTimelineEvent` produces the record shape a `culture-events.tsv` row
 * uses (`culture_profile_id`, `year`, `lane`, `event_type`, `title`, ...), where
 * `year` is the start year — the promotion target for a reviewer.
 */

import type { ContributionSource, Contribution } from "./contribution-service";

// ============================================================================
// Vocabulary & bounds
// ============================================================================

/**
 * Swim-lane keys — kept in sync with the client timeline
 * (`web/src/components/culture-profile/culture-evolution-timeline-utils.ts`
 * `LANE_ORDER`). Redeclared here so the server has no client import.
 */
export const TIMELINE_LANES = [
  "political",
  "territory",
  "urbanism",
  "technology",
  "religion",
  "language",
  "economy",
] as const;
export type TimelineLane = (typeof TIMELINE_LANES)[number];

/** Event magnitude — drives the marker radius on the client timeline. */
export const TIMELINE_MAGNITUDES = ["major", "moderate", "minor"] as const;
export type TimelineMagnitude = (typeof TIMELINE_MAGNITUDES)[number];

/** A single-year marker vs. a dated range. */
export const TIMELINE_ENTRY_KINDS = ["event", "period"] as const;
export type TimelineEntryKind = (typeof TIMELINE_ENTRY_KINDS)[number];

/**
 * Inclusive year bounds an authored entry must fall within (negative = BCE).
 * Guards against fat-fingered years far outside any plausible historical range.
 */
export const TIMELINE_MIN_YEAR = -50000;
export const TIMELINE_MAX_YEAR = 2100;

/** Marks the provenance of an authored timeline contribution. */
export const TIMELINE_PROVENANCE = "user-authored" as const;

// ============================================================================
// Types
// ============================================================================

export interface TimelineEventInput {
  /** `event` = single year; `period` = start..end range. */
  kind: TimelineEntryKind;
  /** The existing culture/entity this entry is associated with (required). */
  cultureProfileId: string;
  /** Human-readable label for the event/period. */
  title: string;
  lane: TimelineLane;
  /** Free-text classifier (e.g. "founding", "conquest"). Defaults to "event". */
  eventType?: string;
  /** Defaults to "moderate". */
  magnitude?: TimelineMagnitude;
  /** Start year (negative = BCE). Required. */
  timePeriodStart: number;
  /**
   * End year. Required for `period`; must be omitted/null (or equal to start)
   * for `event`. Must not precede the start (inverted range).
   */
  timePeriodEnd?: number | null;
  description?: string;
  confidence?: number;
  sources?: ContributionSource[];
  notes?: string;
  contributorName?: string;
  contributorEmail?: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

/** Optional extra bounds (e.g. an associated culture's active period). */
export interface TimelineBounds {
  min: number;
  max: number;
}

// ============================================================================
// Validation
// ============================================================================

function isFiniteNumber(n: unknown): n is number {
  return typeof n === "number" && Number.isFinite(n);
}

function withinBounds(year: number, bounds?: TimelineBounds): boolean {
  const min = Math.max(TIMELINE_MIN_YEAR, bounds?.min ?? TIMELINE_MIN_YEAR);
  const max = Math.min(TIMELINE_MAX_YEAR, bounds?.max ?? TIMELINE_MAX_YEAR);
  return year >= min && year <= max;
}

/**
 * Validate a timeline-event submission: kind, entity association, lane, and a
 * coherent, in-bounds time range. `bounds` optionally tightens the allowed year
 * window (intersected with the global `[TIMELINE_MIN_YEAR, TIMELINE_MAX_YEAR]`).
 */
export function validateTimelineEvent(
  input: Partial<TimelineEventInput>,
  bounds?: TimelineBounds,
): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!input.kind || !TIMELINE_ENTRY_KINDS.includes(input.kind)) {
    errors.push(`kind must be one of: ${TIMELINE_ENTRY_KINDS.join(", ")}`);
  }

  if (!input.title || typeof input.title !== "string" || !input.title.trim()) {
    errors.push("title is required");
  }

  if (
    !input.cultureProfileId ||
    typeof input.cultureProfileId !== "string" ||
    !input.cultureProfileId.trim()
  ) {
    errors.push(
      "cultureProfileId is required — a timeline entry must be associated with an entity",
    );
  }

  if (!input.lane || !TIMELINE_LANES.includes(input.lane)) {
    errors.push(`lane must be one of: ${TIMELINE_LANES.join(", ")}`);
  }

  if (input.magnitude !== undefined && !TIMELINE_MAGNITUDES.includes(input.magnitude)) {
    errors.push(`magnitude must be one of: ${TIMELINE_MAGNITUDES.join(", ")}`);
  }

  // Start year: required, finite, in-bounds.
  if (!isFiniteNumber(input.timePeriodStart)) {
    errors.push("timePeriodStart is required and must be a number (negative = BCE)");
  } else if (!withinBounds(input.timePeriodStart, bounds)) {
    errors.push(
      `timePeriodStart ${input.timePeriodStart} is out of bounds (allowed ${Math.max(
        TIMELINE_MIN_YEAR,
        bounds?.min ?? TIMELINE_MIN_YEAR,
      )}..${Math.min(TIMELINE_MAX_YEAR, bounds?.max ?? TIMELINE_MAX_YEAR)})`,
    );
  }

  // End year depends on kind.
  const hasEnd = input.timePeriodEnd !== undefined && input.timePeriodEnd !== null;
  if (input.kind === "period") {
    if (!hasEnd) {
      errors.push("timePeriodEnd is required for a period entry");
    } else if (!isFiniteNumber(input.timePeriodEnd)) {
      errors.push("timePeriodEnd must be a number");
    } else {
      if (!withinBounds(input.timePeriodEnd, bounds)) {
        errors.push(
          `timePeriodEnd ${input.timePeriodEnd} is out of bounds (allowed ${Math.max(
            TIMELINE_MIN_YEAR,
            bounds?.min ?? TIMELINE_MIN_YEAR,
          )}..${Math.min(TIMELINE_MAX_YEAR, bounds?.max ?? TIMELINE_MAX_YEAR)})`,
        );
      }
      if (isFiniteNumber(input.timePeriodStart) && input.timePeriodEnd < input.timePeriodStart) {
        errors.push("timePeriodEnd must not be earlier than timePeriodStart (inverted range)");
      } else if (
        isFiniteNumber(input.timePeriodStart) &&
        input.timePeriodEnd === input.timePeriodStart
      ) {
        warnings.push("timePeriodEnd equals timePeriodStart; consider kind 'event' for a point in time");
      }
    }
  } else if (input.kind === "event" && hasEnd) {
    if (isFiniteNumber(input.timePeriodEnd) && input.timePeriodEnd !== input.timePeriodStart) {
      errors.push(
        "an event is a single point in time; use kind 'period' for a range (timePeriodEnd must match timePeriodStart)",
      );
    }
  }

  if (input.confidence !== undefined) {
    if (!isFiniteNumber(input.confidence) || input.confidence < 1 || input.confidence > 100) {
      errors.push("confidence must be a number between 1 and 100");
    }
  } else {
    warnings.push("confidence not specified, defaulting to 60");
  }

  return { valid: errors.length === 0, errors, warnings };
}

// ============================================================================
// Serialization
// ============================================================================

/**
 * The canonical `culture-events.tsv` row shape for a timeline entry. `year` is
 * the start year (the single-year column that TSV uses); the full range is kept
 * on the contribution's `entityData` for a reviewer.
 */
export interface SerializedTimelineEvent {
  culture_profile_id: string;
  year: number;
  lane: TimelineLane;
  event_type: string;
  title: string;
  description: string;
  magnitude: TimelineMagnitude;
  sources: string;
}

/**
 * Serialize a timeline entry into the `culture-events.tsv` row shape (with the
 * `sources` cell as the JSON-string-of-titles the loader `JSON.parse`s).
 */
export function serializeTimelineEvent(input: TimelineEventInput): SerializedTimelineEvent {
  const sourceTitles = (input.sources ?? []).map((s) => s.title);
  return {
    culture_profile_id: input.cultureProfileId,
    year: input.timePeriodStart,
    lane: input.lane,
    event_type: input.eventType?.trim() || "event",
    title: input.title,
    description: input.description ?? "",
    magnitude: input.magnitude ?? "moderate",
    sources: JSON.stringify(sourceTitles),
  };
}

// ============================================================================
// Mapping to a contribution
// ============================================================================

/**
 * Map a validated timeline entry onto a `Partial<Contribution>` for the review
 * queue. The full range + provenance live on `entityData`; the serialized
 * `culture-events.tsv` row is mirrored under `entityData.serialized` so a
 * reviewer can promote it verbatim.
 */
export function timelineEventToContribution(input: TimelineEventInput): Partial<Contribution> {
  const confidence = input.confidence ?? 60;
  const isPeriod = input.kind === "period";
  const timePeriodEnd = isPeriod ? (input.timePeriodEnd ?? null) : null;

  const entityData: Record<string, unknown> = {
    kind: input.kind,
    cultureProfileId: input.cultureProfileId,
    title: input.title,
    lane: input.lane,
    eventType: input.eventType?.trim() || "event",
    magnitude: input.magnitude ?? "moderate",
    timePeriodStart: input.timePeriodStart,
    timePeriodEnd,
    description: input.description ?? "",
    source: TIMELINE_PROVENANCE,
    serialized: serializeTimelineEvent(input),
  };

  const sources: ContributionSource[] =
    input.sources && input.sources.length > 0
      ? input.sources
      : [{ title: "User-authored timeline entry" }];

  return {
    entityType: "timeline-event",
    action: "add",
    entityId: input.cultureProfileId,
    entityData,
    sources,
    confidence,
    notes: input.notes,
    contributorName: input.contributorName,
    contributorEmail: input.contributorEmail,
  };
}

File: services/api/src/pinakes/authoring/timeline_event.py
"""Timeline-event authoring — the port of `server/services/timeline-event.ts`.

An `event` (a single year) or a `period` (a dated range) authored on the
temporal axis becomes a queued contribution with provenance
``entityData.source = "user-authored"``; a reviewer promotes it into
`data/source/lexicons/culture-events.tsv` later. Nothing here writes a TSV.

The error strings are the contract, not prose: `post-timeline-event-invalid`
records the **400 body** of an empty submission — message, the five errors in
declaration order, and the one warning — and that fixture is replayed against
both backends.
"""

from __future__ import annotations

from typing import Any, Final, NamedTuple

from pinakes.authoring._js import (
    MISSING,
    field,
    has_sources,
    is_finite_number,
    is_present,
    json_text,
    number_text,
    source_titles,
)

#: Swim-lane keys, kept in sync with the client timeline's ``LANE_ORDER``.
TIMELINE_LANES: Final[tuple[str, ...]] = (
    "political",
    "territory",
    "urbanism",
    "technology",
    "religion",
    "language",
    "economy",
)

#: Event magnitude — drives the marker radius on the client timeline.
TIMELINE_MAGNITUDES: Final[tuple[str, ...]] = ("major", "moderate", "minor")

#: A single-year marker vs. a dated range.
TIMELINE_ENTRY_KINDS: Final[tuple[str, ...]] = ("event", "period")

#: Inclusive year bounds an authored entry must fall within (negative = BCE).
TIMELINE_MIN_YEAR: Final = -50000
TIMELINE_MAX_YEAR: Final = 2100

#: Marks the provenance of an authored timeline contribution.
TIMELINE_PROVENANCE: Final = "user-authored"

#: Confidence a submission gets when it declares none (and warns about it).
DEFAULT_TIMELINE_CONFIDENCE: Final = 60


class ValidationResult(NamedTuple):
    """The verdict, as the route reports it."""

    valid: bool
    errors: list[str]
    warnings: list[str]


def _bounds(bounds: Any) -> tuple[float, float]:
    """The allowed window: the global bounds intersected with *bounds*."""
    low: Any = MISSING
    high: Any = MISSING
    if isinstance(bounds, dict):
        low = bounds.get("min", MISSING)
        high = bounds.get("max", MISSING)
    minimum = max(TIMELINE_MIN_YEAR, low if is_present(low) else TIMELINE_MIN_YEAR)
    maximum = min(TIMELINE_MAX_YEAR, high if is_present(high) else TIMELINE_MAX_YEAR)
    return minimum, maximum


def _within_bounds(year: float, bounds: Any) -> bool:
    minimum, maximum = _bounds(bounds)
    return minimum <= year <= maximum


def _out_of_bounds(label: str, year: Any, bounds: Any) -> str:
    minimum, maximum = _bounds(bounds)
    return (
        f"{label} {number_text(year)} is out of bounds "
        f"(allowed {number_text(minimum)}..{number_text(maximum)})"
    )


def validate_timeline_event(
    body: Any, bounds: Any = None
) -> ValidationResult:
    """Validate a submission: kind, association, lane, and a coherent range.

    ``bounds`` optionally tightens the year window (e.g. an associated culture's
    active period); it is intersected with the global bounds, never widened past
    them.
    """
    errors: list[str] = []
    warnings: list[str] = []

    kind = field(body, "kind")
    if kind not in TIMELINE_ENTRY_KINDS:
        errors.append(f"kind must be one of: {', '.join(TIMELINE_ENTRY_KINDS)}")

    title = field(body, "title")
    if not isinstance(title, str) or not title.strip():
        errors.append("title is required")

    culture_profile_id = field(body, "cultureProfileId")
    if not isinstance(culture_profile_id, str) or not culture_profile_id.strip():
        errors.append(
            "cultureProfileId is required — a timeline entry must be associated "
            "with an entity"
        )

    lane = field(body, "lane")
    if lane not in TIMELINE_LANES:
        errors.append(f"lane must be one of: {', '.join(TIMELINE_LANES)}")

    magnitude = field(body, "magnitude")
    if magnitude is not MISSING and magnitude not in TIMELINE_MAGNITUDES:
        errors.append(f"magnitude must be one of: {', '.join(TIMELINE_MAGNITUDES)}")

    start = field(body, "timePeriodStart")
    if not is_finite_number(start):
        errors.append(
            "timePeriodStart is required and must be a number (negative = BCE)"
        )
    elif not _within_bounds(start, bounds):
        errors.append(_out_of_bounds("timePeriodStart", start, bounds))

    end = field(body, "timePeriodEnd")
    has_end = is_present(end)
    if kind == "period":
        if not has_end:
            errors.append("timePeriodEnd is required for a period entry")
        elif not is_finite_number(end):
            errors.append("timePeriodEnd must be a number")
        else:
            if not _within_bounds(end, bounds):
                errors.append(_out_of_bounds("timePeriodEnd", end, bounds))
            if is_finite_number(start) and end < start:
                errors.append(
                    "timePeriodEnd must not be earlier than timePeriodStart "
                    "(inverted range)"
                )
            elif is_finite_number(start) and end == start:
                warnings.append(
                    "timePeriodEnd equals timePeriodStart; consider kind 'event' "
                    "for a point in time"
                )
    elif kind == "event" and has_end:
        if is_finite_number(end) and end != start:
            errors.append(
                "an event is a single point in time; use kind 'period' for a "
                "range (timePeriodEnd must match timePeriodStart)"
            )

    confidence = field(body, "confidence")
    if confidence is not MISSING:
        if not is_finite_number(confidence) or confidence < 1 or confidence > 100:
            errors.append("confidence must be a number between 1 and 100")
    else:
        warnings.append(
            f"confidence not specified, defaulting to {DEFAULT_TIMELINE_CONFIDENCE}"
        )

    return ValidationResult(valid=not errors, errors=errors, warnings=warnings)


def _event_type(body: Any) -> str:
    """``input.eventType?.trim() || "event"``."""
    raw = field(body, "eventType")
    return raw.strip() if isinstance(raw, str) and raw.strip() else "event"


def _magnitude(body: Any) -> Any:
    raw = field(body, "magnitude")
    return "moderate" if raw is MISSING else raw


def serialize_timeline_event(body: Any) -> dict[str, Any]:
    """The `culture-events.tsv` row shape.

    ``year`` is the *start* year — that TSV carries a single year column, so the
    full range survives only on the contribution's ``entityData`` for a reviewer.
    """
    description = field(body, "description")
    return {
        "culture_profile_id": field(body, "cultureProfileId"),
        "year": field(body, "timePeriodStart"),
        "lane": field(body, "lane"),
        "event_type": _event_type(body),
        "title": field(body, "title"),
        "description": "" if description is MISSING else description,
        "magnitude": _magnitude(body),
        "sources": json_text(source_titles(field(body, "sources"))),
    }


def timeline_event_to_contribution(body: Any) -> dict[str, Any]:
    """Map a validated entry onto the queue's ``Partial<Contribution>`` shape."""
    confidence = field(body, "confidence")
    if confidence is MISSING:
        confidence = DEFAULT_TIMELINE_CONFIDENCE

    kind = field(body, "kind")
    end = field(body, "timePeriodEnd")
    if kind == "period":
        time_period_end = None if end is MISSING else end
    else:
        time_period_end = None

    description = field(body, "description")
    entity_data: dict[str, Any] = {
        "kind": kind,
        "cultureProfileId": field(body, "cultureProfileId"),
        "title": field(body, "title"),
        "lane": field(body, "lane"),
        "eventType": _event_type(body),
        "magnitude": _magnitude(body),
        "timePeriodStart": field(body, "timePeriodStart"),
        "timePeriodEnd": time_period_end,
        "description": "" if description is MISSING else description,
        "source": TIMELINE_PROVENANCE,
        "serialized": serialize_timeline_event(body),
    }

    sources = field(body, "sources")
    contribution: dict[str, Any] = {
        "entityType": "timeline-event",
        "action": "add",
        "entityId": field(body, "cultureProfileId"),
        "entityData": entity_data,
        "sources": (
            sources
            if has_sources(sources)
            else [{"title": "User-authored timeline entry"}]
        ),
        "confidence": confidence,
    }
    # `JSON.stringify` drops an undefined property, and the queue reproduces
    # that (`contributions.store._compact`); an absent optional must not arrive
    # as an explicit null.
    for key in ("notes", "contributorName", "contributorEmail"):
        value = field(body, key)
        if value is not MISSING:
            contribution[key] = value
    return contribution


def options() -> dict[str, Any]:
    """The vocabulary + year bounds the client's authoring form needs."""
    return {
        "kinds": list(TIMELINE_ENTRY_KINDS),
        "lanes": list(TIMELINE_LANES),
        "magnitudes": list(TIMELINE_MAGNITUDES),
        "minYear": TIMELINE_MIN_YEAR,
        "maxYear": TIMELINE_MAX_YEAR,
    }


__all__ = [
    "DEFAULT_TIMELINE_CONFIDENCE",
    "TIMELINE_ENTRY_KINDS",
    "TIMELINE_LANES",
    "TIMELINE_MAGNITUDES",
    "TIMELINE_MAX_YEAR",
    "TIMELINE_MIN_YEAR",
    "TIMELINE_PROVENANCE",
    "ValidationResult",
    "options",
    "serialize_timeline_event",
    "timeline_event_to_contribution",
    "validate_timeline_event",
]
